tests/snap_test.go

```diff
@@ -0,0 +1,201 @@
+	utils.SnapRemove(nil, snapName)
+
+	log.Println("[SETUP]")
+
+	teardown = func() {
+		log.Println("[TEARDOWN]")
+		utils.SnapDumpLogs(nil, start, snapName)
+
+		utils.Exec(nil, "rm "+newPath)
+
+		log.Println("Removing installed snap:", !utils.SkipTeardownRemoval)
+		if !utils.SkipTeardownRemoval {
+			utils.SnapRemove(nil, snapName)
+		}
+	}
+
+	// authorize gpio mock
+	if utils.LocalServiceSnap() {
+		newPath, err = authorizeGpioMock(utils.LocalServiceSnapPath)
+	}
+	if err != nil {
+		teardown()
+		return
+	}
+
+	// install matter-pi-gpio-commander
+	if utils.LocalServiceSnap() {
+		err = utils.SnapInstallFromFile(nil, newPath)
+	} else {
+		err = utils.SnapInstallFromStore(nil, snapName, utils.ServiceChannel)
+	}
+	if err != nil {
+		teardown()
+		return
+	}
+
+	if err = setupGPIO(); err != nil {
+		teardown()
+		return
+	}
+
+	// connect interfaces:
+	utils.SnapConnect(nil, snapName+":avahi-control", "")
+	utils.SnapConnect(nil, snapName+":bluez", "")
+	utils.SnapConnect(nil, snapName+":network", "")
+	utils.SnapConnect(nil, snapName+":network-bind", "")
+	utils.SnapConnect(nil, snapName+":custom-gpio", snapName+":custom-gpio-dev")
+
+	return
+}
+
+func getMockGPIO() (string, error) {
+	gpioChipNumber, stderr, err := utils.Exec(nil, "ls /dev/gpiochip* | sort -n | tail -n 1 | cut -d'p' -f3")
+	if err != nil || stderr != "" {
+		log.Printf("Failed to get mock gpio chip number, Error %s: %s", stderr, err)
+		return "", err
+	}
+	return gpioChipNumber, nil
+}
+
+func setupGPIO() error {
+
+	TestedGpioChip := os.Getenv(specificGpioChip)
+	TestedGpioLine := os.Getenv(specifGpioLine)
+
+	gpiochip := TestedGpioChip
+	var err error
+	if gpiochip == "" {
+		gpiochip, err = getMockGPIO()
+		if err != nil {
+			log.Printf("Failed to get mock gpio chip number: %s", err)
+			return err
+		}
+
+		log.Printf("[TEST] No specific gpiochip defined, using mockup gpio: /dev/gpiochip%s", gpiochip)
+	}
+
+	gpioline := TestedGpioLine
+	if gpioline == "" {
+		gpioline = "4"
+		log.Printf("[TEST] No specific gpio line defined, using default: %s", gpioline)
+	}
+
+	utils.SnapSet(nil, snapName, "gpiochip", gpiochip)
+	utils.SnapSet(nil, snapName, "gpio", gpioline)
+
+	return nil
+}
+
+func TestBlinkOperation(t *testing.T) {
+	log.Println("[TEST] Standard blink operation")
+
+	// test blink operation
+	t.Run("test-blink", func(t *testing.T) {
+		ctx, cancel := context.WithDeadline(context.Background(), <-time.After(20*time.Second))
+		defer cancel()
+
+		stdout, _, err := utils.ExecContextVerbose(nil, ctx, snapName+".test-blink")
+		t.Logf("err: %s", err)
+		t.Logf("stdout: %s", stdout)
+		assert.Error(t, err, "Expected an error")
+		assert.Equal(t, "context deadline exceeded", err.Error())
+	})
+}
+
+func TestWifiCommander(t *testing.T) {
+	log.Println("[TEST] Wifi commander")
+
+	// install chip-tool
+	err := utils.SnapInstallFromStore(t, chipToolSnap, utils.ServiceChannel)
+	if err != nil {
+		t.Fatalf("Failed to install chip-tool: %s", err)
+	}
```

Comment:
The test function has no cleanup implementation before and after: Chip Tool will not get removed after finishing the test, and a second run of the tests on the same environment will have unexpected results.

Reply:
Added cleanup implementation on `setup()` function and on `teardown` declaration for Chip Tool.

Reply:
The cleanup of operations performed in a test should ideally be performed in that test itself. This is for clarity and also to avoid conflict with other future tests relying on Chip Tool. Moreover, each test should be self-contained to allow running it standalone with `go test -run TestMyTest`, which would just run TestMain and TestMyTest.

Reply:
As the `TestMain` run if the test is queried to run individually, so the needed parts to setup the test also run. Thank you.

Reply:
I think my comment has been misunderstood. Each test should be self-contained, and each test should revert the permanent changes that it makes to the host. This is true for TestMain and all the other tests. Keep in mind that Go doesn't guaranty the order of test execution, and the tests can run in any order.

In the given case, another test can be implemented which also needs a fresh installation of Chip Tool, but it could fail depending on the test execution order.

Reply:
Understood, thank you for clarifying. In this particular case, the tests fulfill this requirement. When running only the `TestWifiMatterCommander` test (expected to run after `TestBlinkOperation`), it behaves as expected. It is self-contained and reverts the permanent changes it makes (assuming the environment variable `SKIP_TEARDOWN_REMOVAL` is set to `false`).

While I could include the installation of `chip-tool` within the `setup` phase, it would result in its installation for `TestBlinkOperation` as well, which may not be necessary in this context. What are your thoughts on this approach? It aims to accommodate future test implementations that may require `chip-tool`.

Reply:
As mentioned, the order of tests aren't guaranteed. In this case,  `TestWifiMatterCommander` runs after `TestBlinkOperation` perhaps because tests are run sequentially in alphanumeric order. 

`TestWifiMatterCommander` installs Chip Tool but doesn't remove it. How is that self-contained?

Tomorrow, a contributor can implement another test called `TestThreadMatterCommander` which may run before `TestWifiMatterCommander`. In that situation, `TestWifiMatterCommander` would fail depending on what `TestThreadMatterCommander` has done with Chip Tool. Example of such failure is non-unique matter node IDs or commissioning a device twice. 



> While I could include the installation of `chip-tool` within the `setup` phase, it would result in its installation for `TestBlinkOperation` as well, which may not be necessary in this context. 

This would be another step in the wrong direction, causing more potential inter-test dependency issues.

Reply:
Okay, in that case, now it's fixed! Thank you!